### ISSUE TITLE
Since both rights and resource locations must be disclosed to the AS, the AS can act as Big Brother

### ISSUE THREAD

**Issue by Denisthemalice** [ietf-wg-gnap/gnap-core-protocol#288]
The title says it all. 

This document has been built along “no-privacy by design” rules, despite the fact that it has been said on the mailing list: 
“**_privacy is important_**”. 

At the current time, all the properties that would allow an AS to act as Big Brother have been defined in the document 
(and none to stop it). By design, the AS is able to gather information it needs to act as Big Brother. 

When token introspection is supported by the AS, the AS has all the information it needs to act as Big Brother and even more, since the AS may return to the RS information that was NOT present in the access token !

NOTE: This can be inferred from the following sentence, which is in the draft-ietf-gnap-resource-servers-01 document:

> Structured tokens can also be used in combination with introspection, allowing the token itself to carry one class of information and the introspection response to carry another.

If only capabilities continue to be supported (and not ACLs), the fact that the AS necessarily knows which operations will very likely be performed on every resource and at which point of time should be advertised in the privacy considerations section.

**Comment by jricher** [comment]
This is already discussed in privacy considerations and can be expanded, if you would like to provide some less-inflammatory text.

**Comment by agropper** [comment]
As I read it, the spec allows any RO to specify any AS to the RS or to
request a set of capabilities from the RS. This gives the RO two different
ways avoid surveillance by operating a self-sovereign GNAP AS:
- register their AS with the RO
- share the capabilities with their AS

Please confirm,

- Adrian

**Comment by jricher** [comment]
@agropper That is correct -- not ever RS will accept "any" AS, but it's free to do so and the core document doesn't care how that happens (the RS document has some ways for it to happen). To your second point, the spec allows the processing of attributes/capabilities/policies to be externalized from the AS by allowing the interaction phase to be something other than "log into the AS with your account", as it is assumed to be in OAuth 2. This flexibility allows the RO to present a wide variety of things to the AS in order to make the access token, and these don't have to be identity claims so long as they fulfill the policy that will be ultimately enforced by the RS.

**Comment by Denisthemalice** [comment]
The key sentence I wrote is:

> **_If only capabilities continue to be supported (i.e. and not ACLs)_**, the fact that the AS necessarily knows which operations will very likely be performed on every resource and at which point of time should be advertised in the privacy considerations section.


Since you have all the arguments, you are well placed to write an appropriate text _when capabilities are supported._

However, if attributes only (which imply the use of ACLs)  are supported, **_it is possible to prevent the AS to behave as Big Brother_**.

**Comment by fimbault** [comment]
We discussed RS hiding a while back to avoid disclosing the real location, if needed

**Comment by Denisthemalice** [comment]
**_RS hiding_** can been done in some cases when using attributes, but **_not when using rights_**, since a right 
is composed of an operation and of a resource location. By disclosing the resource location to the AS, 
the AS knows which is the RS.

**Comment by jricher** [comment]
Access can be achieved using mechanisms in the current draft using attributes, what you call rights (which are more properly termed "capabilities" in the security literature), or any number of other combinations. User attributes are not necessary for the non-disclosure principle that you are discussing. This has been demonstrated with actual code.

**Comment by Denisthemalice** [comment]
"Rights" is the wording used in the draft by the editors, while as you noticed, "capabilities" is the wording used in the security literature. 

It would be worthwhile to indicate this relationship in the draft and to have at least one occurrence of the word "capability" 
or "capabilities" in the draft.

You wrote:

> Access can be achieved using mechanisms in the current draft using attributes.

Unfortunately this is not the case with the current draft since section 2.1.1 (Requesting a Single Access Token) states:

> To request a single access token, the client instance sends an "acccess_token" object composed of the following fields.
> 
>    access (array of objects/strings)  Describes the rights that the client instance is requesting for one or more access tokens to be used at RS’s.  This field is **_REQUIRED_**.
> 
> (...)
> 
>    flags (array of strings)  A set of flags that indicate desired attributes or behavior to be attached to the access token by the AS.  This field is **_OPTIONAL_**.
> 

The field "access" is **_REQUIRED_** while the field "flags" is **_OPTIONAL_**. This means that currently the document can only support "capabilities", as called in the security literature.

The field "**_flags_**", as currently described, is unable to support attribute types and attribute values. 
An additional field, called "**_attributes_**" should be added.

Now, let us come back to the title of this thread which is: 
"Since both rights and resource locations must be disclosed to the AS, the AS can act as Big Brother".

The field "access" is REQUIRED which means that the AS will know which kind of actions may be performed on which resource
(which is specified with the field "locations").

You wrote:

> User attributes are not necessary for the non-disclosure principle that you are discussing.

AFAIK, two types of accesses are supposed to be supported by this document: **_EITHER_** capabilities **_OR_** attributes. 
If capabilities are not being used, then attributes must be used.

If only attributes are being used, then the fields "**_access_**" and "**_locations_**" are no more needed.

However, in order to prevent the replay of an access token by a RS towards another RS, another field (e.g. called "**_RS_Id_**" for "**_RS Identifier_**") is needed to indicate an identifier of the RS. A way to hide to the AS the true value of that RS identifier has been discussed and explained on the mailing list about a year ago. Such a field is necessary when attributes only are being used.

When capabilities are being used, such a field should also be used to prevent the replay of an access token by a RS towards another RS (rather than using the field "locations"). 

When attributes are being used, it is unnecessary to disclose to the AS the kind of actions that will be performed on specific resources at the RS. It is also possible in some cases to hide to the AS the true value of that RS identifier. In such a case, the AS cannot act as Big Brother.

Since both "capabilities" and "attributes" have their own merits and drawbacks, both should be supported.

**Comment by jricher** [comment]
Among other problems with this argument, I want to point out that you are conflating the use of the `access` field, which allows the client to describe what it wants, with fields placed inside a structured access token such as an audience restriction or subject identifier, neither of which need to be specified by the client. Since the access token is opaque to the client, these are separately managed. 

Additionally, as has been stated many times previously and discussed at length in other issues such as #244, it is already possible to add an `attributes` field inside of the `access` array if desired by a given RS. Also please recall that the `locations` array is optional for the client to specify.
